alerts/machine-config-operator/MachineConfigControllerDrainError.md

```diff
@@ -0,0 +1,114 @@
+The machine-config-controller pod is the central point of management for incoming updates to 
+machines.
+
+For the commands below replace the $CONTROLLERPOD variable with the name of your own machine-config-controller pod name. 
+
+```console
+oc -n openshift-machine-config-operator logs $CONTROLLERPOD -c machine-config-controller
+```
+
+When a drain is initiated the following log will be recorded.
+
+```console
+1 drain_controller.go:173] node xxxxxxx-xxxxxx-xxxxxx-xxxxx: initiating drain
+```
+The controller will then log each and every pod drained from the node. 
+
+```console
+  1 drain_controller.go:173] node xxxxxxx-xxxxxx-xxxxxx-xxxxx: Evicted pod <namespace>/<pod-name>
+```
+
+If the controller is unable to drain a pod after 1m30s then it will fire an error log.
+
+```console
+1 drain_controller.go:173] node xxxxxxx-xxxxxx-xxxxxx-xxxxx: Drain failed. Waiting 1 minute then retrying. Error message from drain: error when waiting for pod "xxxx-xxxx-xx" in namespace "xxxxxxx" to terminate: global timeout reached: 1m30s
+```
+
+If the drain continues to fail then another failure log will fire.
+
+```console
+ 1 drain_controller.go:173] node xxxxxxx-xxxxxx-xxxxxx-xxxxx: Drain failed. Drain has been failing for more than 10 minutes. Waiting 5 minutes then retrying. Error message from drain: error when waiting for pod "xxxx-xxxx-xx" in namespace "xxxxxxx" to terminate: global timeout reached: 1m30s
+```
+
+Finally if 1 hour has passed and the drain is still failing the following logs will be fired and the node will be marked degraded.
+
+```console
+1 drain_controller.go:352] node xxxxxxx-xxxxxx-xxxxxx-xxxxx: drain exceeded timeout: 1h0m0s. Will continue to retry.
+```
+
+```console
+1 status.go:126] Degraded Machine: xxxxxxx-xxxxxx-xxxxxx-xxxxx: and Degraded Reason: failed to drain node: xxxxxxx-xxxxxx-xxxxxx-xxxxx after 1 hour. Please see machine-config-controller logs for more information
+```
+
+The machine-controller logs will explicitly tell you what pods are failing to drain succesfully. Looking at the pods named in the logs is where you should start. 
+
+Common reasons a node cannot drain a pod successfully are:
+
+- The pod has a PodDisruptionBudget (PBD) set preventing the controller from deleting the pod.
+- The pod has storage attached and the kubelet is unable to unmount the storage succesfully.
+- The pod deletion may be prevented by a webhook if it's configured to target UPDATE operations or the webhook is failing to be called by the kube-apiserver.
+- The pod deletion could be stuck because the finalizers set in the pod are preventing it from terminating succesfully.
+
+## Mitigation 
+
+- If a PDB is causing the failure you can temporarily patch it to set the `minAvailable` to 0 so it can scale the pod down successfully. Then patch it to the old value after the upgrade completes.
```

Comment:
```suggestion
- If a PDB is causing the failure you can temporarily patch it to set the `minAvailable` to 0 so it can scale the pod down successfully. Then patch it to the previous value after the upgrade completes.
```

Reply:
Agreed, Applied the change